File: proposed-projects/round-trip-experiment/reverse/20260909T184722Z_089a4294/20260910T101353Z_4b392f72/ported_implementation/src/rules-builder/symbol-ids.ts
/** We don't need to delete, just preserve relationships. */
export class SymbolIds {
  #map = new Map<string, number>();
  #pos = new Map<string, number>();
  #reverseMap = new Map<number, string>();

  get size(): number {
    return this.#map.size;
  }

  keys(): IterableIterator<string> {
    return this.#map.keys();
  }

  has(key: string): boolean {
    return this.#map.has(key);
  }

  get(key: string): number {
    const value = this.#map.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return value;
  }

  reverseGet(key: number): string {
    const value = this.#reverseMap.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return value;
  }

  getPos(key: string): number {
    const value = this.#pos.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return value;
  }

  set(key: string, value: number, pos: number): void {
    this.#map.set(key, value);
    this.#pos.set(key, pos);
    this.#reverseMap.set(value, key);
  }

  *[Symbol.iterator](): Generator<[string, number]> {
    yield* this.#map.entries();
  }
}

File: packages/gbnf-experiment/data/20260909T193332Z_f85fdfbb/ported_implementation/rules_builder/symbol_ids.py
from __future__ import annotations

from typing import Iterator


class SymbolIdsError(KeyError):
    """Raised when a name or id is missing from a :class:`SymbolIds`."""

    def __str__(self) -> str:
        return str(self.args[0]) if self.args else ""


class SymbolIds:
    """A bidirectional name <-> id map that also remembers where each name was seen."""

    # we don't need to delete, just preserve relationships
    def __init__(self) -> None:
        self._map: dict[str, int] = {}
        self._pos: dict[str, int] = {}
        self._reverse_map: dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self._map)

    def keys(self) -> list[str]:
        return list(self._map.keys())

    def has(self, key: str) -> bool:
        return key in self._map

    def get(self, key: str) -> int:
        if key not in self._map:
            raise SymbolIdsError(f"SymbolIds does not contain key: {key}")
        return self._map[key]

    def reverse_get(self, key: int) -> str:
        if key not in self._reverse_map:
            raise SymbolIdsError(f"SymbolIds does not contain value: {key}")
        return self._reverse_map[key]

    def get_pos(self, key: str) -> int:
        if key not in self._pos:
            raise SymbolIdsError(f"SymbolIds does not contain key: {key}")
        return self._pos[key]

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._pos[key] = pos
        self._reverse_map[value] = key

    def __contains__(self, key: str) -> bool:
        return self.has(key)

    def __len__(self) -> int:
        return self.size

    def __iter__(self) -> Iterator[tuple[str, int]]:
        yield from self._map.items()

    # camelCase aliases, mirroring the reference API.
    reverseGet = reverse_get
    getPos = get_pos
